src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -1297,14 +1297,10 @@ private static object DoUncheckedConversion(SpecialType destinationType, Constan
                         }
                     case ConstantValueTypeDiscriminator.Single:
                     case ConstantValueTypeDiscriminator.Double:
-                        // This code used to invoke CheckConstantBounds and return constant zero if the value is not within the target type.
-                        // The C# spec says that in this case the result of the conversion is an unspecified value of the destination type.
-                        // Zero is a perfectly valid unspecified value, so that behavior was formally correct.
-                        // But it did not agree with the behavior of the native C# compiler, that apparently returned a value that
-                        // would resulted from a runtime conversion with normal CLR overflow behavior.
-                        // To avoid breaking programs that might accidentally rely on that unspecified behavior
-                        // we now removed that check and just allow conversion to overflow.
-                        double doubleValue = value.DoubleValue;
+                        // When converting from a floating-point type to an integral type, if the checked conversion would
+                        // throw an overflow exception, then the unchecked conversion is undefined.  So that we have
+                        // identical behavior on every host platform, we yield a result of zero in that case.
+                        double doubleValue = CheckConstantBounds(destinationType, value.DoubleValue) ? value.DoubleValue : 0D;
```

Comment:
>doubleValue [](start = 31, length = 11)

Reverting [https://github.com/dotnet/roslyn/commit/c121bc53e4cadd98a1beda3c902e0521c833eaa8](https://github.com/dotnet/roslyn/commit/c121bc53e4cadd98a1beda3c902e0521c833eaa8)

Reply:
It would be good to dig on bug 1020273. Did someone report the difference in behavior between roslyn and native compiler? Should we document this change even though we're only changing an unspecified behavior?

---
In reply to: [311307945](https://github.com/dotnet/roslyn/pull/37772#discussion_r311307945) [](ancestors = 311307945)

Reply:
Yes, a customer did report a difference between the native compiler and Roslyn.  However, the "fixed" behavior still differs depending on what platform you run the compiler on (as I suspect the native compiler also differed among platforms).  This PR ensures we get consistent (deterministic) behavior on all host platforms.